openslides_backend/action/motion/set_number_mixin.py

```diff
@@ -0,0 +1,124 @@
+            ["motions_number_type", "motions_number_min_digits"],
+        )
+        if meeting.get("motions_number_type") == "manually":
+            return
+        state = self.datastore.get(
+            FullQualifiedId(Collection("motion_state"), state_id), ["set_number"]
+        )
+        if not state.get("set_number"):
+            return
+
+        # Generate the components of the number
+        prefix = self.get_prefix(meeting_id, lead_motion_id, category_id)
+        number_value = self.get_number_value(
+            meeting_id, lead_motion_id, category_id, existing_number_value
+        )
+
+        # Generate test number and check uniqueness
+        number_value_str = str(number_value).rjust(
+            meeting.get("motions_number_min_digits", 0), "0"
+        )
+        number = f"{prefix}{number_value_str}"
+        while not self.check_if_unique(number, meeting_id):
+            number_value += 1
+            number_value_str = str(number_value).rjust(
+                meeting.get("motions_number_min_digits", 0), "0"
+            )
+            number = f"{prefix}{number_value_str}"
+
+        instance["number"] = number
+        instance["number_value"] = number_value
+
+    def get_prefix(
+        self, meeting_id: int, lead_motion_id: Optional[int], category_id: Optional[int]
+    ) -> str:
+        meeting = self.datastore.get(
+            FullQualifiedId(Collection("meeting"), meeting_id),
+            ["motions_number_with_blank", "motions_amendments_prefix"],
+        )
+        blank = " " if meeting.get("motions_number_with_blank") else ""
+        if lead_motion_id:
+            lead_motion = self.datastore.get(
+                FullQualifiedId(Collection("motion"), lead_motion_id), ["number"]
+            )
+            prefix = f"{lead_motion.get('number', '')}{blank}{meeting.get('motions_amendments_prefix', '')}"
+        elif not category_id:
+            prefix = ""
+        else:
+            category = self.datastore.get(
+                FullQualifiedId(Collection("motion_category"), category_id), ["prefix"]
+            )
+            if category.get("prefix"):
+                prefix = f"{category['prefix']}{blank}"
+            else:
+                prefix = ""
+        return prefix
+
+    def get_number_value(
+        self,
+        meeting_id: int,
+        lead_motion_id: Optional[int],
+        category_id: Optional[int],
+        existing_number_value: Optional[int],
+    ) -> int:
+        meeting = self.datastore.get(
+            FullQualifiedId(Collection("meeting"), meeting_id), ["motions_number_type"]
+        )
+        if existing_number_value:
+            return existing_number_value
+
+        if lead_motion_id:
+            filter: Union[And, FilterOperator] = FilterOperator(
+                "lead_motion_id", "=", lead_motion_id
+            )
+        elif meeting.get("motions_number_type") == "per_category":
+            filter = FilterOperator("category_id", "=", category_id)
```

Comment:
here, a filter for the meeting is missing: If the category id is `None`, you want only motions from the current meeting.

Reply:
Done.